FWCore/Framework/src/EventProcessor.cc

```diff
@@ -1780,130 +2047,202 @@ namespace edm {
+                  // handleNextEvent takes care of immediately ending the run properly
+                  // using the same code it uses to end the run in other situations.
+                }
+                handleNextEventForStreamAsync(std::move(iTask), iStreamIndex);
+              });
 
           processEventAsync(WaitingTaskHolder(*iTask.group(), recursionTask), iStreamIndex);
         } else {
-          //the stream will stop now
-          if (status->isLumiEnding()) {
-            if (lastTransitionType() == InputSource::IsLumi and not status->haveStartedNextLumi()) {
-              status->startNextLumi();
-              beginLumiAsync(status->nextSyncValue(), status->runResource(), iTask);
+          // the stream will stop processing this lumi now
+          if (status->eventProcessingState() == LuminosityBlockProcessingStatus::kStopLumi) {
+            if (not status->haveStartedNextLumiOrEndedRun()) {
+              status->startNextLumiOrEndRun();
+              if (lastTransitionType() == InputSource::IsLumi && !iTask.taskHasFailed()) {
+                CMS_SA_ALLOW try {
+                  beginLumiAsync(IOVSyncValue(EventID(input_->run(), input_->luminosityBlock(), 0),
+                                              input_->luminosityBlockAuxiliary()->beginTime()),
+                                 streamRunStatus_[iStreamIndex],
+                                 iTask);
+                } catch (...) {
+                  WaitingTaskHolder copyHolder(iTask);
+                  copyHolder.doneWaiting(std::current_exception());
+                  endRunAsync(streamRunStatus_[iStreamIndex], iTask);
+                }
+              } else {
+                // If appropriate, this will also start the next run.
+                endRunAsync(streamRunStatus_[iStreamIndex], iTask);
+              }
             }
             streamEndLumiAsync(std::move(iTask), iStreamIndex);
           } else {
-            iTask.doneWaiting(std::exception_ptr{});
+            assert(status->eventProcessingState() == LuminosityBlockProcessingStatus::kPauseForFileTransition);
+            auto runStatus = streamRunStatus_[iStreamIndex].get();
+
+            if (runStatus->holderOfTaskInProcessRuns().hasTask()) {
+              runStatus->holderOfTaskInProcessRuns().doneWaiting(std::exception_ptr{});
+            }
           }
         }
       } catch (...) {
-        // It is unlikely we will ever get in here ...
-        // But if we do try to clean up and propagate the exception
-        if (streamLumiStatus_[iStreamIndex]) {
-          streamEndLumiAsync(iTask, iStreamIndex);
-        }
-        bool expected = false;
-        if (deferredExceptionPtrIsSet_.compare_exchange_strong(expected, true)) {
-          auto e = std::current_exception();
-          deferredExceptionPtr_ = e;
-          iTask.doneWaiting(e);
-        }
+        WaitingTaskHolder copyHolder(iTask);
```

Comment:
I believe there could be a possible use-after-move here. The `iTask` is moved-from on line 2232 to an argument of `streamEndLumiAsync()`, and if that function throws an exception (I did not look how likely that would be), `iTask` is in after-move state here.

Reply:
Done. I removed the std::move. Probably the performance gain of that move is negligible anyway. That resolves the main issue in your comment. The other question in the comment is harder.

There is a part of me that wants things to be absolutely perfect, but sometimes they are not. Can streamEndLumiAsync throw? Probably in 100 years it will not throw once. But there is a call to make_waiting_task in there that does a tiny memory allocation and also potentially setting up a task chain, which again might do a tiny memory allocation. So streamEndLumiAsync could throw. If it does throw, even with this move gone I think the program will probably get stuck and fail. But it is very very very unlikely to throw. Note that the calls to the module end stream functions are enclosed in try-catch blocks and such exceptions would all be caught and handled properly. It is really hard to deal with exceptions in code that is somewhat similar to a destructor, it is cleaning up and releasing resources... If the resources are not released the program can get stuck... It bothers me, but I think it is good enough. This problem was already there before this PR...

If we were building a nuclear reactor, I would say it must be fixed...

For it to throw, the program would have to be running out of memory already and this particular tiny memory allocations out of millions of such allocations would have to be the one that failed...